README.md
# rollout-redis ⛳️

[![Gem Version](https://badge.fury.io/rb/rollout-redis.svg)](https://badge.fury.io/rb/rollout-redis)

Fast and easy feature flags based on Redis. 

Based on the discontinued [rollout](https://github.com/fetlife/rollout) project, removing some capabilities, including some new features and supporting latest Redis versions.

Topics covered in this README:

- [Install it](#install-it)
- [Quick Start](#quick-start-)
- [Advanced features](#advanced-features-)
  - [Gradual activation based on percentages](#gradual-activation-based-on-percentages)
  - [Caching Feature Flags](#caching-feature-flags)
  - [Auto-deactivating flags](#auto-deactivating-flags)
  - [Sending Notifications](#sending-notifications)
- [Rake tasks](#rake-tasks)
- [Migrating from rollout gem](#migrating-from-rollout-gem-)
  - [Compatible payloads and keys](#compatible-payloads-and-keys)
- [Changelog](#changelog)
- [Contributing](#contributing)

## Install it 

```bash
gem install rollout-redis
```

## Quick Start 💨

Instantiate the `Rollout` class sending a `Redis` instance as a parameter.

```ruby
require 'redis'
require 'rollout'

@redis ||= Redis.new(
        host: ENV.fetch('REDIS_HOST'),
        port: ENV.fetch('REDIS_PORT')
      )
@rollout ||= Rollout.new(@redis)
```

Now you can activate Feature Flags:

```ruby
@rollout.activate('FEATURE_FLAG_NAME') # => true/false
```

Verify if a feature is currently enabled:

```ruby
if @rollout.active?('FEATURE_FLAG_NAME')
  # your new code here...
end
```

An alternative to the if check, is to wrap your code under the `with_feature` method. The wrapped code will be performed only if the feature flag is active:

```ruby
@rollout.with_feature('FEATURE_FLAG_NAME') do
  # your new code here...
end
```

If there is an issue, you have the option to disable a feature:

```ruby
@rollout.deactivate('FEATURE_FLAG_NAME')
```

If you want to list all the stored feature flags, you can use the `features` method:

```ruby
@rollout.features
```

The response will be an array of hashes with all the information about the stored feature flags

```ruby
[
  { name: 'a-feature-flag', percentage: 100, data: { requests: 50, errors: 1 } },
  { name: 'another-feature-flag', percentage: 20, data: { requests: 1, errors: 0 } },
  { name: 'super-feature-flag', percentage: 50, data: { requests: 828, errors: 34 } }
]
```

## Advanced features 🦾

### Gradual activation based on percentages

When introducing a new feature, it's a recommended practice to gradually enable it for a specific portion of your target audience to evaluate its impact. To achieve this, you can utilize the `activate` method, as shown below:

```ruby
@rollout.activate('FEATURE_FLAG_NAME', 20)
```

Now, to know if a feature flags is enabled, you need to provide a determinator (in this example, we're using the user email):

```ruby
if @rollout.active?('FEATURE_FLAG_NAME', user_email)
  # your new code here...
end
```

The gradual activation also works wrapping your code within the `with_feature` method, you just need to provde the determinator you want to use.

```ruby
@rollout.with_feature('FEATURE_FLAG_NAME', user_email) do
  # your new code here...
end
```

It's important to note that if you use the `active?` method without specifying a determinator to determine whether this subset of the audience should see the new feature, it will always return `false` since the activation percentage is less than 100%. See:

```ruby
@rollout.activate('FEATURE_FLAG_NAME', 20)
@rollout.active?('FEATURE_FLAG_NAME') # => false
```

### Caching Feature Flags

The Rollout gem is tightly integrated with Redis for feature flag status management. Consequently, occasional connectivity issues between your application and the Redis storage may arise.

To prevent potential application degradation when the Redis storage is unavailable, you can enable feature flag status caching during the gem's instantiation:

```ruby
@rollout ||= Rollout.new(redis).with_cache
```

Additionally, you can specify extra parameters to configure the duration (in seconds) for which the feature flag status is stored in the cache. By default, this duration is set to 300 seconds (5 minutes):

```ruby
@rollout ||= Rollout.new(redis)
              .with_cache(expires_in: 300)
```

In the case that you need to clear the cache at any point, you can make use of the `clean_cache` method:

```ruby
@rollout.clean_cache
```

### Auto-deactivating flags

If you want to allow the gem to deactivate all the feature flags automatically when a threshold of errors is reached, you can enable the degrade feature using the `with_degrade` method.

```ruby
@rollout ||= Rollout.new(redis)
              .with_cache
              .with_degrade(min: 100, threshold: 0.1)
```

However, if you just want to activate the degradation of an specific feature flag, you need to provide the following information when activating the feature flag (note that now the percentage is a mandatory parameter if you want to pass the degrade options):

```ruby
@rollout.activate('FEATURE_FLAG_NAME', 100, degrade: { min: 500, threshold: 0.2 })
```

The same configuration 👆 is available when using the rake task for activating the feature flag. Check [Rake tasks](#rake-tasks) section.

```shell
bundle exec rake rollout:on[FEATURE_FLAG_NAME,100,500,0.2]
```

So now, instead of using the `active?` method, you need to wrap your new code under the `with_feature` method.

```ruby
@rollout.with_feature('FEATURE_FLAG_NAME') do
  # your new feature code here...
end
```

When any unexpected error appears during the wrapped code execution, the Rollout gem will take it into account for automatically degrading the feature flag if the threshold of errors is reached. The feature flag will not be removed from the redis, but it will change its percentage to 0 and it will be marked as degraded.

_NOTE_: All the managed or captured errors inside the wrapped code will not be taken into consideration for degrading the feature flag.

### Sending notifications

`rollout-redis` gem can send different notifications to your development team. For enabling this feature, you just need to use the `with_notifications` instance method providing the channels where you want to publish each of the different events that can occur:

- **status_change**: This notification is triggered when a feature flag is activated or deactivated using the `rollout-redis` gem.
- **degrade**: This notification is triggered when a feature flag is automatically degraded because the threshold of errors is reached
  - The instance must be configured for automatically degrading using the `with_degrade` instance method.

You must provide at least one [channel](#defining-the-channels) as a parameter if you want to enable the notifications for that specific event. If no channels provided, the notifications will not be sent. 

```ruby
@rollout ||= Rollout.new(redis)
              .with_cache
              .with_degrade(min: 100, threshold: 0.1)
              .with_notifications(
                status_change: [slack_channel],
                degrade: [slack_channel, email_channel]
              )
```

#### Defining the channels

When enabling a notification, you can provide the different channels where the notification should be published. `rollout-redis` gem offers different channels that can be configured.

##### Slack Channel

Allows you to send notifications using a slack webhook.

The first thing to do is to setup an incoming webhook service integration. You can do this from your services page.

After that, you can provide the obtained webhook url when instantiating the Slack channel.

```ruby
require 'rollout'

slack_channel = Rollout::Notifications::Channels::Slack.new(
  webhook_url: ENV.fetch('SLACK_COMPANY_WEBHOOK_URL'),
  channel: '#feature-flags-notifications',
  username: 'rollout-redis'
)
```

##### Email Channel

Allows you to send notifications via email.

```ruby
require 'rollout'

email_channel = Rollout::Notifications::Channels::Email.new(
  smtp_host: ENV.fetch('SMTP_HOST'),
  smtp_port: ENV.fetch('SMTP_PORT'),
  from: 'no-reply@rollout-redis.com',
  to: 'developers@yourcompany.com'
)
```

##### Custom channel

If you want to send the notifications using a different channel not offered by this gem, you can implement your own class with a `publish` method.

```ruby
require 'rollout'

module YourApp
  class YourCustomChannel
    def initialize()
      # provide here whatever you need for configuring your channel
    end

    def publish(text)
      # Implement the way you want to publish the notification
    end
  end
end
```

After implementing it, you can pass it to the list of configured channels

```ruby
your_channel = YourApp::YourCustomChannel.new(url: 'wadus', param2: 'foo')
@rollout ||= Rollout.new(redis)
              .with_cache
              .with_degrade(min: 100, threshold: 0.1)
              .with_notifications(
                status_change: [your_channel],
                degrade: [your_channel]
              )
```

## Rake tasks

In order to have access to the rollout rakes, you have to load manually the task definitions. For doing so load the rollout rake task:

```ruby
require 'rollout'

load 'rollout/tasks/rollout.rake'
```

Also, for using the rake tasks, you must set the following env variables

```shell
ROLLOUT_REDIS_HOST=localhost
ROLLOUT_REDIS_PORT=6379
```

### Usage

To activate/deactivate features, execute the following rake tasks:

```shell
bundle exec rake rollout:on[feature_name]
bundle exec rake rollout:off[feature_name]
```

To a gradual activation based on percentages, pass the percentage as the second parameter when executing the `on` task.

```shell
bundle exec rake rollout:on[feature_name,50]
```

_NOTE_: In both cases, `feature_name` **must not** include quotes e.g. `bundle exec rake rollout:on['feature_name']`, as the gem will be unable to fetch its activation status if so.

For listing all the stored feature flags, do:

```shell
bundle exec rake rollout:list
```

For migrating feature flags stored using the old `rollout` gem format (check [migration guide](https://github.com/jcagarcia/rollout-redis/blob/main/MIGRATING_FROM_ROLLOUT_GEM.md)), do:

```shell
bundle exec rake rollout:migrate_from_rollout_format
```

## Migrating from rollout gem 🚨

If you are currently using the unmaintained [rollout](https://github.com/fetlife/rollout) gem, you should consider checking this [migration guide](https://github.com/jcagarcia/rollout-redis/blob/main/MIGRATING_FROM_ROLLOUT_GEM.md) for start using the new `rollout-redis` gem.

### Compatible payloads and keys

You can use the `.with_old_rollout_gem_compatibility` instance method for making the `rollout-redis` gem work as the the discontinued [rollout](https://github.com/fetlife/rollout) gem in terms of redis storage and format storage. Check the [migration guide](https://github.com/jcagarcia/rollout-redis/blob/main/MIGRATING_FROM_ROLLOUT_GEM.md) for more information.

## Changelog

If you're interested in seeing the changes and bug fixes between each version of `rollout-redis`, read the [Changelog](https://github.com/jcagarcia/rollout-redis/blob/main/CHANGELOG.md).

## Contributing

We welcome and appreciate contributions from the open-source community. Before you get started, please take a moment to review the guidelines below.

### How to Contribute

1. Fork the repository.
2. Clone the repository to your local machine.
3. Create a new branch for your contribution.
4. Make your changes and ensure they meet project standards.
5. Commit your changes with clear messages.
6. Push your branch to your GitHub repository.
7. Open a pull request in our repository.
8. Participate in code review and address feedback.
9. Once approved, your changes will be merged.

### Development

This project is dockerized, so be sure you have docker installed in your machine. 

Once you clone the repository, you can use the `Make` commands to build the project.

```shell
make build
```

You can pass the tests running:

```shell
make test
```

### Issue Tracker

Open issues on the GitHub issue tracker with clear information.

### Contributors

*   Juan Carlos García - Creator - https://github.com/jcagarcia

The `rollout-redis` gem is based on the discontinued [rollout](https://github.com/fetlife/rollout) project, created by [James Golick](https://github.com/jamesgolick)



CHANGELOG.md
# Changelog
All changes to `rollout-redis` will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2023-10-28

### Added
- `#with_old_rollout_gem_compatibility` method for allowing working with Feature Flags that were stored by the old `rollout` gem.
- `rollout:migrate_from_rollout_format` rake task for performing a migration of the feature flags stored by the old `rollout` gem to the new `rollout-redis` format.
- Add a new parameter when performing `#activate` method for providing a specific degrade configuration for the feature flag that is being activated.
- Add a new parameter when performing `rollout:on` rake task for providing a specific degrade configuration for the feature flag that is being activated.
- You can implement now your own `Rollout::Notifications::Channels::Channel` in case the ones offered by the `gem` are not enough.

## [1.0.0] - 2023-10-25

### Added
- `#with_notifications` method for allowing to send notifications when some different event occurs.

### Changed
- When the threshold of errors is reached when using the `with_degrade` feature, instead of deleting the feature flag from the redis, we are marking it now as degraded, moving the activation percentage to 0% and adding some useful information to the feature flag stored data.

## [0.3.1] - 2023-10-24
- Same as 0.3.0. When testing GitHub actions for moving to first release `1.0.0` it deployed a new version of the gem by error.

## [0.3.0] - 2023-10-24

### Added
- Providing some rake tasks to the consumers of the gem for allowing them to easily manage their feature flags in their applications:
    - `bundle exec rake rollout:on` rake task for activating feature flags
    - `bundle exec rake rollout:off` rake task for deactivating feature flags
    - `bundle exec rake rollout:list` rake task for listing stored feature flags

## [0.2.0] - 2023-10-23

### Added

- `#features` method for listing all the feature flags stored in Redis

## [0.1.0] - 2023-10-23

- Initial version


spec/active_spec.rb
require 'spec_helper'

RSpec.describe Rollout do
  class CustomManagedError < StandardError; end

  let(:storage) { MockRedis.new }
  let(:instance) { described_class.new(storage)}
  let(:feature_flag_name) { 'A_FEATURE_FLAG_NAME'}

  after(:each) do
    instance.deactivate(feature_flag_name)
  end

  describe '#active?' do
    context 'when feature flag is fully active' do
      before do
        instance.activate(feature_flag_name)
      end

      it 'returns true' do
        result = instance.active?(feature_flag_name)
  
        expect(result).to be true
      end
    end

    context 'when feature flag is only active for a portion of the target' do
      before do
        instance.activate(feature_flag_name, 20)
      end

      it 'return true for the configured portion of the requests' do
        expect((1..100).count do |count|
          instance.active?(feature_flag_name, "user#{count}@email.com")
        end).to be_within(2).of(20)
      end

      it 'returns false for at least the other portion of the requests' do
        expect((1..100).count do |count|
          !instance.active?(feature_flag_name, "user#{count}@email.com")
        end).to be_within(2).of(80)
      end
    end

    context 'when feature flag is NOT active' do
      it 'returns false' do
        result = instance.active?(feature_flag_name)
  
        expect(result).to be false
      end
    end

    context 'when global degrade is enable' do
      let(:instance_with_degrade) { instance.with_degrade }

      context 'and the feature flag is enabled' do
        before do
          instance.activate(feature_flag_name)
        end

        it 'stores the number of requests' do
          33.times do
            instance_with_degrade.active?(feature_flag_name)
          end

          data = storage.get(key(feature_flag_name))

          expect(data).to_not be nil
          expect(JSON.parse(data, symbolize_names: true)).to eq({
            percentage: 100,
            requests: 33
          })
        end
      end
    end

    context 'when degrade is enable for the specific feature flag' do
      context 'and the feature flag is enabled' do
        let(:instance) { described_class.new(storage) }
        before do
          instance.activate(feature_flag_name, 100, degrade: { min: 100, threshold: 0.5 })
        end

        it 'stores the number of requests' do
          101.times do
            instance.active?(feature_flag_name)
          end

          data = storage.get(key(feature_flag_name))

          expect(data).to_not be nil
          expect(JSON.parse(data, symbolize_names: true)).to eq({
            percentage: 100,
            requests: 101,
            degrade: {
              min: 100,
              threshold: 0.5
            }
          })
        end
      end
    end

    context 'when redis not available' do
      context 'and cache NOT enabled' do
        it 'raises a Rollout::Error' do
          allow(storage).to receive(:get).and_raise(::Redis::BaseError)

          expect {
            instance.active?(feature_flag_name)
          }.to raise_error(Rollout::Error)
        end
      end

      context 'and cache enabled' do
        let(:instance_with_cache) { instance.with_cache }

        context 'and the feature flag was cached before' do
          before do
            instance_with_cache.activate(feature_flag_name)
          end

          context 'and not expired' do
            context 'and the feature flag was fully enabled' do
              it 'returns true' do
                allow(storage).to receive(:get).and_raise(::Redis::BaseError)
                result = instance_with_cache.active?(feature_flag_name)

                expect(result).to be true
              end
            end

            context 'and the feature flag was only active for a portion of the target' do
              before do
                instance_with_cache.activate(feature_flag_name, 20)
              end

              it 'return true for the configured portion of the requests' do
                allow(storage).to receive(:get).and_raise(::Redis::BaseError)
                expect((1..100).count do |count|
                  instance_with_cache.active?(feature_flag_name, "user#{count}@email.com")
                end).to be_within(2).of(20)
              end
        
              it 'returns false for at least the other portion of the requests' do
                allow(storage).to receive(:get).and_raise(::Redis::BaseError)
                expect((1..100).count do |count|
                  !instance_with_cache.active?(feature_flag_name, "user#{count}@email.com")
                end).to be_within(2).of(80)
              end
            end
          end

          context 'but expired' do
            let(:instance_with_cache) { instance.with_cache(expires_in: 1) }

            before do
              allow(Time).to receive(:now).and_return(Time.now + 2)
            end

            it 'raises a Rollout::Error' do
              allow(storage).to receive(:get).and_raise(::Redis::BaseError)

              expect {
                instance_with_cache.active?(feature_flag_name)
              }.to raise_error(Rollout::Error)
            end
          end
        end
      end
    end
  end

  private

  def key(feature_name)
    "feature-rollout-redis:#{feature_name}"
  end
end

spec/spec_helper.rb
require 'mock_redis'
require 'rake'
require 'rollout'

RSpec.configure do |config|
  config.expect_with :rspec do |c|
      c.syntax = :expect
  end

  config.order = :random
end

lib/rollout/feature.rb
# frozen_string_literal: true

require 'zlib'

class Rollout
    class Feature
      attr_accessor :percentage, :degrade
      attr_reader :name, :data

      RAND_BASE = (2**32 - 1) / 100.0
  
      def initialize(name, data={})
        @name = name
        @data = data
        @percentage = @data[:percentage]
        @degrade = @data[:degrade]
      end
  
      def active?(determinator=nil)
        if determinator
          determinator_in_percentage?(determinator)
        else
          @percentage == 100
        end
      end

      def add_request
        if @data[:requests]
          @data[:requests] = @data[:requests] + 1
        else
          @data[:requests] = 1
        end
      end

      def add_error
        if @data[:errors]
          @data[:errors] = @data[:errors] + 1
        else
          @data[:errors] = 1
        end
      end

      def requests
        @data[:requests] || 0
      end

      def errors
        @data[:errors] || 0
      end

      def to_h
        h = {
          name: @name,
          percentage: @percentage,
          data: {
            requests: requests,
            errors: errors
          }
        }

        h.merge!({
          degrade: @degrade
        }) if @degrade

        h
      end
  
      private
  
      def determinator_in_percentage?(determinator)
        Zlib.crc32(determinator) < RAND_BASE * @percentage
      end
    end
  end

lib/rollout/notifications/channels/slack.rb
require 'slack-notifier'

class Rollout
  module Notifications
    module Channels
      class Slack
        def initialize(webhook_url:, channel:, username:'rollout-redis')
          @webhook_url = webhook_url
          @channel = channel
          @username = username
        end

        def publish(text)
          begin
            blocks = [
              {
                "type": "section",
                "text": {
                  "type": "mrkdwn",
                  "text": text
                }
              }
            ]
            slack_notifier.post(blocks: blocks)
          rescue => e
            puts "[ERROR] Error sending notification to slack webhook. Error => #{e}"
          end
        end

        private

        def slack_notifier
          @notifier ||= ::Slack::Notifier.new @webhook_url do
            defaults channel: @channel,
                     username: @username
          end
        end
      end
    end
  end
end

lib/rollout/notifications/notifiers/degrade.rb
require_relative 'base'

class Rollout
  module Notifications
    module Notifiers
      class Degrade < Base
        def initialize(channels)
          super(channels)
        end

        def notify(feature_name, requests, errors)
          @channels.each do |c|
            if c.class == Rollout::Notifications::Channels::Email
              publish_to_email_channel(c, feature_name, requests, errors)
            else
              publish_to_channel(c, feature_name, requests, errors)
            end
          end
        end

        private

        def publish_to_channel(c, feature_name, requests, errors)
          text = "Feature flag '#{feature_name}' has been degraded after #{requests} requests and #{errors} errors"
          c.publish(text)
        end

        def publish_to_email_channel(c, feature_name, requests, errors)
          subject = 'Feature flag has been automatically deactivated!'
          content = "Feature flag '#{feature_name}' has been degraded after #{requests} requests and #{errors} errors"
          c.publish(subject, content)
        end
      end
    end
  end
end

lib/rollout/notifications/notifiers/status_change.rb
require_relative 'base'

class Rollout
  module Notifications
    module Notifiers
      class StatusChange < Base
        def initialize(channels)
          super(channels)
        end

        def notify(feature_name, new_status, new_percentage=nil)
          @channels.each do |c|
            if c.class == Rollout::Notifications::Channels::Email
              publish_to_email_channel(c, feature_name, new_status, new_percentage)
            else
              publish_to_channel(c, feature_name, new_status, new_percentage)
            end
          end
        end

        private

        def publish_to_channel(c, feature_name, new_status, new_percentage)
          if new_status == :activated
            text = "Feature flag '#{feature_name}' has been activated with percentage #{new_percentage}!"
          elsif new_status == :deactivated
            text = "Feature flag '#{feature_name}' has been deactivated and deleted!"
          end
          c.publish(text)
        end

        def publish_to_email_channel(c, feature_name, new_status, new_percentage)
          if new_status == :activated
            subject = 'Feature flag has been activated!'
            content = "Feature flag '#{feature_name}' has been activated with percentage #{new_percentage}!"
          elsif new_status == :deactivated
            subject = 'Feature flag has been deactivated!'
            content = "Feature flag '#{feature_name}' has been deactivated and deleted!"
          end
          c.publish(subject, content)
        end
      end
    end
  end
end

lib/rollout.rb
# frozen_string_literal: true

require 'redis'
require 'json'

require 'rollout/feature'
require 'rollout/notifications/channels/email'
require 'rollout/notifications/channels/slack'
require 'rollout/notifications/notifiers/degrade'
require 'rollout/notifications/notifiers/status_change'
require 'rollout/version'


class Rollout

  class Error < StandardError; end

  attr_reader :storage

  def initialize(storage)
    @storage = storage
    @cache_enabled = false
    @degrade_enabled = false
    @old_gem_compatibility_enabled = false
    @auto_migrate_from_old_format = false
  end

  def with_cache(expires_in: 300)
    @cache_enabled = true
    @cache_time = expires_in
    @cache = {}

    self
  end

  def with_degrade(min: 100, threshold: 0.1)
    @degrade_enabled = true
    @degrade_min = min
    @degrade_threshold = threshold

    self
  end

  def with_notifications(status_change:[], degrade:[])
    status_change_channels = status_change
    degrade_channels = degrade

    if !status_change_channels.empty?
      status_change_channels.each do |c|
        raise Rollout::Error.new("Channel #{c.class.name} does not implement `publish` method") unless c.respond_to?(:publish)
      end
      @status_change_notifier = Notifications::Notifiers::StatusChange.new(status_change_channels)
    end

    if !degrade_channels.empty?
      degrade_channels.each do |c|
        raise Rollout::Error.new("Channel #{c.class.name} does not implement `publish` method") unless c.respond_to?(:publish)
      end
      @degrade_notifier = Notifications::Notifiers::Degrade.new(degrade_channels)
    end

    self
  end

  def with_old_rollout_gem_compatibility(auto_migration: false)
    @old_gem_compatibility_enabled = true
    @auto_migrate_from_old_format = auto_migration

    self
  end

  def activate(feature_name, percentage=100, degrade: nil)
    data = { percentage: percentage }
    data.merge!({
      degrade: {
        min: degrade[:min] || 0,
        threshold: degrade[:threshold] || 0
      }
    }) if degrade

    feature = Feature.new(feature_name, data)
    result = save(feature) == "OK"

    if result
      @cache[feature_name] = {
        feature: feature,
        timestamp: Time.now.to_i
      } if @cache_enabled

      @status_change_notifier&.notify(feature_name, :activated, percentage)
    end

    result
  end

  def activate_percentage(feature_name, percentage)
    activate(feature_name, percentage)
  end

  def deactivate(feature_name)
    result = del(feature_name)

    @status_change_notifier&.notify(feature_name, :deactivated)

    result
  end

  def active?(feature_name, determinator = nil)
    feature = get(feature_name)
    if feature.nil? && @old_gem_compatibility_enabled
      feature = get_with_old_format(feature_name)
      if feature && @auto_migrate_from_old_format
        activate(feature_name, feature.percentage)
      end
    end

    return false unless feature

    active = feature.active?(determinator)

    if active && degrade_enabled?(feature)
      feature.add_request
      save(feature)
    end

    active
  end

  def with_feature_flag(feature_name, determinator = nil, &block)
    yield if active?(feature_name, determinator)
  rescue Rollout::Error => e
    raise
  rescue => e
    feature = get(feature_name)
    if feature && degrade_enabled?(feature)
      feature.add_error
      save(feature)

      degrade(feature_name) if degraded?(feature)
    end
    raise e
  end

  def features
    keys = @storage.keys("#{key_prefix}:*")
    return [] if keys.empty?

    keys.map do |key|
      data = @storage.get(key)
      next unless data

      feature_name = key.gsub("#{key_prefix}:", '')
      Feature.new(feature_name, JSON.parse(data, symbolize_names: true)).to_h
    end
  end

  def clean_cache
    return unless @cache_enabled

    @cache = {}
  end

  def migrate_from_rollout_format
    keys = @storage.keys('feature:*')

    keys.each do |old_key|
      new_key = old_key.gsub('feature:', 'feature-rollout-redis:')
      old_data = @storage.get(old_key)

      if old_data
        percentage = old_data.split('|')[0].to_i

        new_data = {
          percentage: percentage,
          requests: 0,
          errors: 0
        }.to_json

        @storage.set(new_key, new_data)

        puts "Migrated redis key from #{old_key} to #{new_key}. Migrating data from '#{old_data}' to '#{new_data}'."

        if percentage > 0
          @status_change_notifier&.notify(new_key.gsub('feature-rollout-redis:', ''), :activated, percentage)
        end
      end
    end
  end

  private

  def get(feature_name)
    feature = from_redis(feature_name)
    return unless feature

    @cache[feature_name] = {
      feature: feature,
      timestamp: Time.now.to_i
    } if @cache_enabled

    feature
  rescue ::Redis::BaseError => e
    cached_feature = from_cache(feature_name)
    raise Rollout::Error.new(e) unless cached_feature

    cached_feature
  end

  def get_with_old_format(feature_name)
    feature = from_redis_with_old_format(feature_name)
    return unless feature

    feature
  rescue ::Redis::BaseError => e
    raise Rollout::Error.new(e)
  end

  def save(feature)
    @storage.set(key(feature.name), feature.data.to_json)
  end
  
  def del(feature_name)
    @storage.del(key(feature_name)) == 1
  end

  def degrade(feature_name)
    feature = get(feature_name)
    data_with_degrade = feature.data.merge({
      percentage: 0,
      degraded: true,
      degraded_at: Time.now
    })
    result = @storage.set(key(feature.name), data_with_degrade.to_json) == "OK"

    if result
      @degrade_notifier.notify(feature_name, feature.requests, feature.errors)
    end

    result
  end

  def from_cache(feature_name)
    return nil unless @cache_enabled

    cached = @cache[feature_name]

    if expired?(cached[:timestamp])
      @cache.delete(feature_name)
      return nil
    end

    cached[:feature]
  end

  def from_redis(feature_name)
    data = @storage.get(key(feature_name))
    return unless data
    Feature.new(feature_name, JSON.parse(data, symbolize_names: true))
  end

  def from_redis_with_old_format(feature_name)
    old_data = @storage.get(old_key(feature_name))
    return unless old_data

    percentage = old_data.split('|')[0].to_i

    new_data = {
      percentage: percentage,
      requests: 0,
      errors: 0
    }

    Feature.new(feature_name, new_data)
  end

  def expired?(timestamp)
    Time.now.to_i - timestamp > @cache_time
  end

  def degraded?(feature)
    return false if !degrade_enabled?(feature)

    if feature.degrade
      degrade_min = feature.degrade[:min]
      degrade_threshold = feature.degrade[:threshold]
    else
      degrade_min = @degrade_min
      degrade_threshold = @degrade_threshold
    end

    return false if feature.requests < degrade_min

    feature.errors > degrade_threshold * feature.requests
  end

  def degrade_enabled?(feature)
    @degrade_enabled || !feature.degrade.nil?
  end

  def key(name)
    "#{key_prefix}:#{name}"
  end

  def old_key(name)
    "#{old_key_prefix}:#{name}"
  end

  def key_prefix
    "feature-rollout-redis"
  end

  def old_key_prefix
    "feature"
  end
end